packages/menu/src/Menu.ts

```diff
@@ -201,30 +370,90 @@ export class Menu extends SpectrumElement {
             item = this.menuItems[index] as MenuItem;
         }
         index = Math.max(index, 0);
+        const selectedItemsMap = new Map<MenuItem, boolean>();
+        const selected: string[] = [];
+        const selectedItems: MenuItem[] = [];
         this.menuItems.forEach((item, i) => {
+            if (item.selected) {
+                selectedItemsMap.set(item, true);
+                selected.push(item.value);
+                selectedItems.push(item);
+            }
             if (i !== index) {
                 item.focused = false;
             }
         });
+        this.selectedItemsMap = selectedItemsMap;
+        this.selected = selected;
+        this.selectedItems = selectedItems;
         this.focusedItemIndex = index;
         this.focusInItemIndex = index;
     }
 
-    private prepItems = (): void => {
-        this.menuItems = [
-            ...this.querySelectorAll(`[role="${this.childRole}"]`),
-        ] as MenuItem[];
-        if (!this.menuItems || this.menuItems.length === 0) {
-            return;
+    // debounce update this.menuItems so they're in DOM order
+    private async updateMenuItems(): Promise<void> {
+        this.itemsChanged = true;
+
+        return new Promise((resolve) => {
+            // debounce so we only actually update once
+            requestAnimationFrame(() => {
+                if (!this.itemsChanged) {
+                    resolve();
+                }
+
+                this.menuItems = [];
+
+                const slotElements =
+                    this.menuSlot.assignedElements({ flatten: true }) || [];
+                for (const slotElement of slotElements) {
+                    const childItems: MenuItem[] =
+                        slotElement instanceof MenuItem
+                            ? [slotElement as MenuItem]
+                            : ([
+                                  ...slotElement.querySelectorAll(
+                                      `[role="${this.childRole}"]`
+                                  ),
+                              ] as MenuItem[]);
+                    for (const childItem of childItems) {
+                        if (this.menuItemSet.has(childItem)) {
+                            this.menuItems.push(childItem);
+                        }
+                    }
+                }
+
+                this.updateSelectedItemIndex();
```

Comment:
Possibly nit as O(1n) and O(2n) are kind of the same...

This walks the array we just built again, is there room to merge the work?

Reply:
there might be, but there are times we only need to do updateSelectedItemIndex. I'll take a look once I get more of the changes here cleaned up.